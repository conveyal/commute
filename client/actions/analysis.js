import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  const commuters = newAnalysis.commuters
  delete newAnalysis.commuters
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`),
    makeMockTrips(newAnalysis.id, organizationId, commuters)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organizationId) => [
  deleteLocally({ analysisId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]

const makeMockTrips = (analysisId, organizationId, commuters) => {
  const trips = []
  const modes = ['bike', 'car', 'transit', 'walk']
  const randRange = () => Math.random() * 0.4 + 0.8
  commuters.forEach((commuter) => {
    const curTrip = {
      commuterId: commuter.id
    }
    const tripMetricBase = Math.random()
    let bestValue = 9999999
    let bestMode
    modes.forEach((mode) => {
      // generate random trips
      const modePctMultiplier = randRange()
      const cost = tripMetricBase * 34.7 * modePctMultiplier * randRange()
      const distance = tripMetricBase * 69 * modePctMultiplier * randRange()
      const time = tripMetricBase * 7500 * modePctMultiplier * randRange()
      curTrip[mode] = {cost, distance, time}
      const tripValue = cost + time * 0.01
      if (tripValue < bestValue) bestMode = mode
    })
    curTrip.mostLikely = Object.assign({ mode: bestMode }, curTrip[bestMode])
    trips.push(curTrip)
  })
  return sendMockTrips({analysisId, organizationId, trips})
}
const sendMockTrips = createAction('receive mock calculated trips')
