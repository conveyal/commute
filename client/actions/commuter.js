import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addCommuter = createAction('add commuter')
export const createCommuter = ({commuter, groupId, organizationId}) => {
  commuter.id = uuid.v4()
  return [
    addCommuter({commuter, groupId, organizationId}),
    push(`/organizations/${organizationId}/groups/${groupId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete commuter')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteCommuter = ({commuterId, groupId, organizationId}) => [
  deleteLocally({commuterId, groupId, organizationId}),
  push(`/organizations/${organizationId}/groups/${groupId}`)
]

const updateLocally = createAction('update commuter')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateCommuter = ({commuter, groupId, organizationId}) => [
  updateLocally({commuter, groupId, organizationId}),
  push(`/organizations/${organizationId}/groups/${groupId}`)
]
