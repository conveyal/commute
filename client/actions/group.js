import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addGroup = createAction('add group')
const appendCommuters = createAction('append commuters')
export const appendToGroup = (newCommuters, groupId, organizationId) => {
  return [
    appendCommuters({ commuters: newCommuters, groupId, organizationId }),
    push(`/organizations/${organizationId}/groups/${groupId}`)
  ] // TODO save to server
}
export const createGroup = (newGroup, organizationId) => {
  newGroup.id = uuid.v4()
  newGroup.allAddressesGeocoded = false
  return [
    addGroup({ group: newGroup, organizationId }),
    push(`/organizations/${organizationId}/groups/${newGroup.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete group')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/group/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteGroup = (id, organizationId) => [
  deleteLocally({ groupId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]
