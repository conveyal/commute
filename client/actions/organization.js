import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add organization')
export const createOrganization = (newOrganization) => {
  newOrganization.id = uuid.v4()
  newOrganization.analyses = []
  newOrganization.analysesById = {}
  newOrganization.sites = []
  newOrganization.sitesById = {}
  newOrganization.groups = []
  newOrganization.groupsById = {}
  return [
    addLocally(newOrganization),
    push(`/organizations/${newOrganization.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete organization')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/organization/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteOrganization = (id) => [deleteLocally(id), push('/')]

export const updateOrganization = (organization) => [
  setLocally(organization),
  push(`/organizations/${organization.id}`)
]
/* const updateOnServer = (organization) =>
  serverAction({
    url: `/api/organization/${organization.id}`,
    options: {
      body: JSON.stringify(organization),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set organization')
