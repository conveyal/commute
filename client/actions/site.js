import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addSite = createAction('add site')
export const createSite = (newSite, organizationId) => {
  newSite.id = uuid.v4()
  return [
    addSite({ organizationId, site: newSite }),
    push(`/organizations/${organizationId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete site')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/site/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteSite = (id, organizationId) => [
  deleteLocally({ siteId: id, organizationId }),
  push(`/organizations/${organizationId}`)
]

const updateLocally = createAction('update site')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/site/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateSite = (site, organizationId) => [
  updateLocally({ organizationId, site }),
  push(`/organizations/${organizationId}`)
]
