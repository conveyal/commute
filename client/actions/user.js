import {createAction} from 'redux-actions'

// login / logout
export const login = createAction('log in')
export const logout = createAction('log out')
