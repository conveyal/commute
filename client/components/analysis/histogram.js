import React, {Component, PropTypes} from 'react'
import {Col, ControlLabel, FormControl, FormGroup, Grid, Panel, Row} from 'react-bootstrap'
import {
  Crosshair,
  DiscreteColorLegend,
  HorizontalGridLines,
  VerticalBarSeries,
  XAxis,
  YAxis
} from 'react-vis'

import ButtonLink from '../button-link'
import FlexiblePlot from '../flexible-plot'
import Icon from '../icon'
import analysisDefaults from '../../utils/analysisDefaults'
import {calcNumLessThan, getInitialSeries} from '../../utils/components'

const {metrics} = analysisDefaults
const metricsVals = Object.keys(metrics)

export default class Histogram extends Component {
  static propTypes = {
    // dispatch
    deleteAnalysis: PropTypes.func,

    // props
    analysis: PropTypes.object.isRequired,
    groupName: PropTypes.string.isRequired,
    site: PropTypes.object.isRequired
  }

  state = {
    crosshairValues: [],
    metric: 'time',
    series: getInitialSeries()
  }

  componentWillMount () {
    // calculate with selected metric
    this._calculateSeries({...this.state})
  }

  /**
   * Calculate series values given selected metric
   *
   * @param  {Object} newState The new state to work with
   */
  _calculateSeries (newState) {
    const arrayVals = this.props.analysis.tripVals
    const metricData = metrics[newState.metric]
    const multiplier = metricData.multiplier || 1

    // reset series data
    newState.series = getInitialSeries()

    // disallow transit distance metric
    if (newState.metric === 'distance') {
      newState.series = newState.series.filter((seriesMode) => seriesMode.mode !== 'transit')
    }

    newState.series = newState.series.map((seriesMode) => {
      seriesMode.data = []
      seriesMode.numInPreviousBins = 0
      return seriesMode
    })

    // iterate through all bins of histogram
    for (let i = 0; i < metricData.max; i += metricData.step) {
      const upper = i + metricData.step
      const curBinLabel = `${i} - ${upper}`

      // calculate number of trips by mode that fall inside current bin
      newState.series.forEach((seriesMode) => {
        // find total num to left
        const numLess = calcNumLessThan(arrayVals[seriesMode.mode][newState.metric],
          upper * multiplier)

        // add bin
        seriesMode.data.push({ x: curBinLabel, y: numLess - seriesMode.numInPreviousBins })

        // update for next iteration
        seriesMode.numInPreviousBins = numLess
      })
    }

    // add those off edge
    newState.series.forEach((seriesMode) => {
      // add bin with remnant
      seriesMode.data.push({
        x: `> ${metricData.max}`,
        y: arrayVals[seriesMode.mode][newState.metric].length - seriesMode.numInPreviousBins
      })
    })

    this.setState(newState)
  }

  /**
   * Format the title line of the crosshair.
   * @param {Array} values Array of values.
   * @returns {Object} The caption and the value of the title.
   * @private
   */
  _formatCrosshairTitle = (values) => {
    return {
      title: metrics[this.state.metric].unit,
      value: values[0].x
    }
  }

  /**
   * A callback to format the crosshair items.
   * @param {Object} values Array of values.
   * @returns {Array<Object>} Array of objects with titles and values.
   * @private
   */
  _formatCrosshairItems = (values) => {
    const {series} = this.state
    return values.map((v, i) => {
      return {
        title: series[i].title,
        value: v.y
      }
    })
  }

  /**
   * Handle the change of metric
   *
   * @param  {[type]} e [description]
   * @private
   */
  _handleMetricChange = (event) => {
    const newState = {...this.state}
    newState.metric = event.target.value
    this._calculateSeries(newState)
  }

  /**
   * Click handler for the legend.
   * @param {Object} item Clicked item of the legend.
   * @param {number} i Index of the legend.
   * @private
   */
  _legendClickHandler = (item, i) => {
    const {series} = this.state
    series[i].disabled = !series[i].disabled
    this.setState({series})
  }

  /**
   * Event handler for onMouseLeave.
   * @private
   */
  _mouseLeaveHandler = () => {
    this.setState({crosshairValues: []})
  }

  /**
   * Event handler for onNearestX.
   * @param {Object} value Selected value.
   * @param {number} index Index of the series.
   * @private
   */
  _nearestXHandler = (value, {index}) => {
    const {series} = this.state
    this.setState({
      crosshairValues: series.map(s => s.data[index])
    })
  }

  render () {
    const {_id: analysisId, name} = this.props.analysis
    const {groupName, site} = this.props
    const {crosshairValues, metric: selectedMetric, series} = this.state
    return (
      <Grid>
        <Row className='histogram-header'>
          <Col xs={12}>
            <h3>
              <span>{name}</span>
              <ButtonLink
                className='pull-right'
                to={`/analysis/${analysisId}`}
                >
                <Icon type='arrow-left' />
                <span>Back</span>
              </ButtonLink>
            </h3>
            <h3>Commute Metrics Histogram</h3>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Site:</strong>
              <span>{site.name}</span>
            </h4>
          </Col>
          <Col xs={12} md={6}>
            <h4>
              <strong>Group:</strong>
              <span>{groupName}</span>
            </h4>
          </Col>
        </Row>
        <Row>
          <Col xs={9}>
            <FlexiblePlot
              animation
              height={300}
              margin={{bottom: 70}}
              onMouseLeave={this._mouseLeaveHandler}
              xType='ordinal'
              >
              <Crosshair
                itemsFormat={this._formatCrosshairItems}
                titleFormat={this._formatCrosshairTitle}
                values={crosshairValues}
                />
              <HorizontalGridLines />
              {series
                .filter((s) => !s.disabled)
                .map((s, idx) =>
                  <VerticalBarSeries
                    color={s.color}
                    data={s.data}
                    key={`possibilities-series-${s.title}`}
                    onNearestX={this._nearestXHandler}
                    />
                )
              }
              <XAxis
                tickLabelAngle={-45}
                title={metrics[selectedMetric].unit}
                />
              <YAxis title='Number of Trips' />
            </FlexiblePlot>
          </Col>
          <Col xs={3}>
            <DiscreteColorLegend
              items={series}
              onItemClick={this._legendClickHandler}
              />
          </Col>
        </Row>
        <Row>
          <Col xs={12} md={8} className='histogram-settings'>
            <h4>Settings</h4>
            <Panel>
              <FormGroup controlId='histogram-metric-select'>
                <ControlLabel>Metric</ControlLabel>
                <FormControl
                  componentClass='select'
                  onChange={this._handleMetricChange}
                  value={selectedMetric}
                  >
                  {metricsVals.map((metric) => {
                    return <option value={metric}>{metrics[metric].label}</option>
                  })}
                </FormControl>
              </FormGroup>
            </Panel>
          </Col>
        </Row>
      </Grid>
    )
  }
}
