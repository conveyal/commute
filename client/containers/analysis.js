import {connect} from 'react-redux'

import {deleteAnalysis} from '../actions/analysis'
import _histogram from '../components/analysis/histogram'
import _individuals from '../components/analysis/individuals'
import _possibilities from '../components/analysis/possibilities'
import _summary from '../components/analysis/summary'

function mapStateToProps (state, props) {
  const {organization} = state
  const {params} = props
  const currentOrganizationId = params.organizationId
  const currentOrganization = organization.organizationsById[currentOrganizationId]
  const curAnalysis = currentOrganization.analysesById[params.analysisId]
  const group = currentOrganization.groupsById[curAnalysis.groupId]
  return {
    analysis: curAnalysis,
    commutersById: group.commutersById,
    groupName: group.name,
    organizationId: currentOrganizationId,
    siteName: currentOrganization.sitesById[curAnalysis.siteId].name
  }
}

function mapDispatchToProps (dispatch) {
  return {
    deleteAnalysis: (id, organizationId) => dispatch(deleteAnalysis(id, organizationId))
  }
}

export const Histogram = connect(mapStateToProps)(_histogram)
export const Individuals = connect(mapStateToProps)(_individuals)
export const Possibilities = connect(mapStateToProps)(_possibilities)
export const Summary = connect(mapStateToProps, mapDispatchToProps)(_summary)
