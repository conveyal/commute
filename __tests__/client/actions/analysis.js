/* global describe, it */

import {expectCreateAnalysis, expectDeleteAnalysis} from '../../test-utils/actions'

import * as analysis from '../../../client/actions/analysis'
import {mockCommuter} from '../../test-utils/mock-data'

describe('actions > analysis', () => {
  it('create analysis should work', () => {
    const data = {
      commuters: [mockCommuter],
      groupId: '1',
      name: 'Mock Analysis',
      siteId: '1'
    }
    const result = analysis.createAnalysis(data, '1')

    expectCreateAnalysis(result)
  })

  it('delete analysis should work', () => {
    const result = analysis.deleteAnalysis('1', '1')

    expectDeleteAnalysis(result)
  })
})
