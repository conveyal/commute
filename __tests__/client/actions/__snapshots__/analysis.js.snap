exports[`actions > analysis create analysis should work 1`] = `
Object {
  "payload": Object {
    "analysis": Object {
      "groupId": "1",
      "name": "Mock Analysis",
      "siteId": "1",
    },
    "organizationId": "1",
  },
  "type": "add analysis",
}
`;

exports[`actions > analysis create analysis should work 2`] = `
Object {
  "payload": Object {
    "args": Array [
      "/organizations/1/analysis/new-entity-id",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;

exports[`actions > analysis delete analysis should work 1`] = `
Array [
  Object {
    "payload": Object {
      "analysisId": "1",
      "organizationId": "1",
    },
    "type": "delete analysis",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/organizations/1",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;
